sdk/tests/deqp/functional/gles3/es3fOcclusionQueryTests.js

```diff
@@ -0,0 +1,461 @@
+
+                bufferedLogToConsole('Occluder draw ' + (i + 1) + ' / ' + this.m_numOccluderDraws + ' : ' + 'Stencil clear');
+                bufferedLogToConsole('Stencil-clearing box drawn at ' +
+                                                '(' + scissorBoxX + ', ' + scissorBoxY + ')' +
+                                                ', width = ' + scissorBoxW + ', height = ' + scissorBoxH + '.');
+
+                gl.enable(gl.SCISSOR_TEST);
+                gl.scissor(scissorBoxX, scissorBoxY, scissorBoxW, scissorBoxH);
+                gl.clearStencil(1);
+                gl.clearColor(STENCIL_CLEAR_COLOR[0], STENCIL_CLEAR_COLOR[1], STENCIL_CLEAR_COLOR[2], STENCIL_CLEAR_COLOR[3]);
+                gl.clear(gl.COLOR_BUFFER_BIT | gl.STENCIL_BUFFER_BIT);
+                gl.disable(gl.SCISSOR_TEST);
+
+                break;
+            }
+
+            default:
+                throw new Error('Invalid occluder type: ' + type);
+        }
+    }
+
+    if (this.m_occluderTypes & OCCLUDER_SCISSOR) {
+        var scissorBoxX = this.m_rnd.getInt(-SCISSOR_OFFSET, targetW - SCISSOR_OFFSET);
+        var scissorBoxY = this.m_rnd.getInt(-SCISSOR_OFFSET, targetH - SCISSOR_OFFSET);
+        var scissorBoxW = this.m_rnd.getInt(SCISSOR_MINSIZE, targetW + SCISSOR_OFFSET);
+        var scissorBoxH = this.m_rnd.getInt(SCISSOR_MINSIZE, targetH + SCISSOR_OFFSET);
+
+        bufferedLogToConsole('Scissor box drawn at ' +
+                                        '(' + scissorBoxX + ', ' + scissorBoxY + ')' +
+                                        ', width = ' + scissorBoxW + ', height = ' + scissorBoxH + '.');
+
+        gl.enable(gl.SCISSOR_TEST);
+        gl.scissor(scissorBoxX, scissorBoxY, scissorBoxW, scissorBoxH);
+    }
+
+    var query = gl.createQuery();
+    gl.beginQuery(this.m_queryMode, query);
+
+    // Draw target primitives
+
+    gl.enable(gl.DEPTH_TEST);
+    gl.enable(gl.STENCIL_TEST);
+    gl.stencilFunc(gl.EQUAL, 0, 0xFF);
+
+    for (var i = 0; i < this.m_numTargetDraws; i++) {
+        var targetVertices = this.generateVertices(2, 2, this.m_numTargetsPerDraw, 3, this.m_rnd, this.m_targetSize, 0.4, 1); // Generate vertices for target primitives
+
+        if (targetVertices.length > 0) {
+            gl.uniform4f(colorUnif, TARGET_COLOR[0], TARGET_COLOR[1], TARGET_COLOR[2], TARGET_COLOR[3]);
+            gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(targetVertices), gl.STATIC_DRAW);
+            gl.drawArrays(gl.TRIANGLES, 0, 3 * this.m_numTargetsPerDraw);
+        }
+    }
+
+    gl.endQuery(this.m_queryMode);
+    gl.finish();
+    gl.disable(gl.SCISSOR_TEST);
+    gl.disable(gl.STENCIL_TEST);
+    gl.disable(gl.DEPTH_TEST);
+
+    // Check that query result is available.
+    var resultAvailable = /** @type {boolean} */ (gl.getQueryParameter(query, gl.QUERY_RESULT_AVAILABLE));
+    assertMsgOptions(resultAvailable, 'Occlusion query failed to return a result after glFinish()', false, true);
```

Comment:
If occlusion query results are not available immediately after glFinish(), then doesn't it make occlusion query useless?
My understanding is that, in real life applications, we use occlusion query with a simple bonding box to see if an object is visible. If it is visible, then we start the real rendering.
If we need to return the control to the browser to get the query results, wouldn't it be too slow to be worth the effort?